File: python/src/basic_math/sieve_of_eratosthenes.py
from typing import List

def sieve_of_eratosthenes(n: int) -> List[int]:
    """
    エラトステネスのふるいを用いて、n以下の素数のリストを返します。
    
    Args:
        n (int): 上限値
        
    Returns:
        List[int]: n以下の素数のリスト
    """
    if n < 2:
        return []
        
    is_prime = [True] * (n + 1)
    is_prime[0] = False
    is_prime[1] = False
    
    p = 2
    while (p * p <= n):
        if is_prime[p]:
            for i in range(p * p, n + 1, p):
                is_prime[i] = False
        p += 1
        
    primes = []
    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)
            
    return primes

File: javascript/src/math/sieve_of_eratosthenes.js
/**
 * エラトステネスのふるいを用いて、n以下の素数のリストを返します。
 * 
 * @param {number} n 上限値
 * @returns {Array} n以下の素数のリスト
 */
function sieveOfEratosthenes(n) {
    if (n < 2) {
        return [];
    }

    const isPrime = new Array(n + 1).fill(true);
    isPrime[0] = false;
    isPrime[1] = false;

    for (let p = 2; p * p <= n; p++) {
        if (isPrime[p]) {
            for (let i = p * p; i <= n; i += p) {
                isPrime[i] = false;
            }
        }
    }

    const primes = [];
    for (let i = 2; i <= n; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }

    return primes;
}

module.exports = { sieveOfEratosthenes };
